// @nationaldesignstudio/rampart/dist/index.js
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/types.ts
var KEEP_LABELS = new Set(["CITY", "STATE", "ZIP_CODE"]);
function resolveKeepLabels(keepLabels) {
  return keepLabels === undefined ? KEEP_LABELS : new Set(keepLabels);
}
function shouldRedact(label, keepLabels = KEEP_LABELS) {
  return !keepLabels.has(label);
}

// src/policy.ts
function mergeSpans(spans) {
  const sorted = [...spans].sort((a, b) => a.start - b.start || b.end - a.end);
  const merged = [];
  for (const span of sorted) {
    const prev = merged[merged.length - 1];
    if (prev === undefined || span.start >= prev.end) {
      merged.push(span);
      continue;
    }
    const winner = preferred(prev, span);
    const prevContains = prev.start <= span.start && prev.end >= span.end;
    const spanContains = span.start <= prev.start && span.end >= prev.end;
    if (prevContains || spanContains) {
      merged[merged.length - 1] = winner;
    } else {
      const start = Math.min(prev.start, span.start);
      const end = Math.max(prev.end, span.end);
      merged[merged.length - 1] = { ...winner, start, end, text: winner.text };
    }
  }
  return merged;
}
function preferred(a, b) {
  if (a.score !== b.score)
    return a.score > b.score ? a : b;
  const aLen = a.end - a.start;
  const bLen = b.end - b.start;
  if (aLen !== bLen)
    return aLen > bLen ? a : b;
  return a.source === "heuristic" ? a : b;
}
function applyPolicy(spans, keepLabels = KEEP_LABELS) {
  return mergeSpans(spans).filter((s) => shouldRedact(s.label, keepLabels)).sort((a, b) => b.start - a.start);
}

// src/ner/classifier.ts
var GROUP_TO_LABEL = {
  GIVEN_NAME: "GIVEN_NAME",
  GIVENNAME: "GIVEN_NAME",
  SURNAME: "SURNAME",
  LASTNAME: "SURNAME",
  EMAIL: "EMAIL",
  PHONE: "PHONE",
  URL: "URL",
  TAX_ID: "TAX_ID",
  BANK_ACCOUNT: "BANK_ACCOUNT",
  ROUTING_NUMBER: "ROUTING_NUMBER",
  GOVERNMENT_ID: "GOVERNMENT_ID",
  PASSPORT: "PASSPORT",
  DRIVERS_LICENSE: "DRIVERS_LICENSE",
  BUILDING_NUMBER: "BUILDING_NUMBER",
  STREET_NAME: "STREET_NAME",
  SECONDARY_ADDRESS: "SECONDARY_ADDRESS",
  SECADDRESS: "SECONDARY_ADDRESS",
  CITY: "CITY",
  STATE: "STATE",
  ZIP_CODE: "ZIP_CODE"
};
var RAMPART_MODEL_ID = "nationaldesignstudio/rampart";
var DEFAULT_OPTIONS = {
  device: "wasm",
  minScore: 0.4
};
var MODEL_MAX_TOKENS = 512;
var SPECIAL_TOKENS = 2;
var NER_TOKEN_BUDGET = MODEL_MAX_TOKENS - SPECIAL_TOKENS - 10;
var NER_TOKEN_OVERLAP = 64;
var COMBINING_MARKS_RE = /\p{M}/gu;
var EXTEND_SCORE = 0.15;
var CONNECTOR_RE = /^[\s'\u2019.-]*$/;
var PERSON_LABELS = new Set(["GIVEN_NAME", "SURNAME"]);
var LEFT_PARTICLE_RE = /([\p{Lu}][\p{L}\p{M}\u2019']{0,3})([\s'\u2019.-]{1,3})$/u;
var RIGHT_PARTICLE_RE = /^([\s'\u2019.-]{1,3})([\p{Lu}][\p{L}\p{M}\u2019']{0,3})/u;
async function loadNerClassifier(options = {}) {
  const { pipeline } = await import("@huggingface/transformers");
  const merged = { ...DEFAULT_OPTIONS, ...options };
  const model = merged.model ?? RAMPART_MODEL_ID;
  const classifier = await pipeline("token-classification", model, {
    dtype: "q4",
    device: merged.device
  });
  const adapter = (text, opts) => classifier(text, opts);
  const tokenizer = classifier.tokenizer;
  if (tokenizer?.encode) {
    adapter.countTokens = (text) => tokenizer.encode(text, { add_special_tokens: false }).length;
  }
  if (tokenizer?.tokenize) {
    adapter.tokenize = (text) => tokenizer.tokenize(text);
  }
  return adapter;
}
async function detectNer(raw, classifier, minScore = DEFAULT_OPTIONS.minScore) {
  const windows = classifier.countTokens === undefined ? [{ start: 0, end: raw.length }] : planTokenWindows(raw, classifier.countTokens, NER_TOKEN_BUDGET, NER_TOKEN_OVERLAP);
  if (windows.length <= 1) {
    return detectNerWindow(raw, classifier, minScore);
  }
  const spans = [];
  for (const window of windows) {
    const windowSpans = await detectNerWindow(raw.slice(window.start, window.end), classifier, minScore);
    for (const span of windowSpans) {
      spans.push({ ...span, start: span.start + window.start, end: span.end + window.start });
    }
  }
  return mergeSpans(spans);
}
function planTokenWindows(raw, countTokens, budget, overlap) {
  const segments = toSegments(raw, countTokens, budget);
  if (segments.length === 0)
    return [];
  const windows = [];
  let i = 0;
  while (i < segments.length) {
    let tokens = 0;
    let j = i;
    while (j < segments.length && (j === i || tokens + segments[j].tokens <= budget)) {
      tokens += segments[j].tokens;
      j++;
    }
    windows.push({ start: segments[i].start, end: segments[j - 1].end });
    if (j === segments.length)
      break;
    let shared = 0;
    let next = j;
    while (next > i + 1 && shared < overlap) {
      next--;
      shared += segments[next].tokens;
    }
    i = next;
  }
  return windows;
}
function toSegments(raw, countTokens, budget) {
  const segments = [];
  for (const [start, end] of wordSpans(raw)) {
    let from = start;
    while (from < end) {
      const tokens = countTokens(raw.slice(from, end));
      if (tokens <= budget) {
        segments.push({ start: from, end, tokens });
        break;
      }
      const cut = fitCharsToBudget(raw, from, end, budget, countTokens);
      segments.push({ start: from, end: cut, tokens: countTokens(raw.slice(from, cut)) });
      from = cut;
    }
  }
  return segments;
}
function* wordSpans(raw) {
  const n = raw.length;
  let i = 0;
  while (i < n) {
    let j = i;
    while (j < n && !/\s/.test(raw[j]))
      j++;
    while (j < n && /\s/.test(raw[j]))
      j++;
    yield [i, j];
    i = j;
  }
}
function fitCharsToBudget(raw, from, end, budget, countTokens) {
  let lo = from + 1;
  let hi = end;
  let best = from + 1;
  while (lo <= hi) {
    const mid = lo + hi >> 1;
    if (countTokens(raw.slice(from, mid)) <= budget) {
      best = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return best;
}
async function detectNerWindow(raw, classifier, minScore = DEFAULT_OPTIONS.minScore) {
  const inferText = raw.replaceAll("-", " ");
  const entities = await classifier(inferText, { aggregation_strategy: "simple" });
  const folded = foldForModel(inferText);
  const indexOffsets = classifier.tokenize === undefined ? undefined : buildTokenIndexOffsets(folded, classifier.tokenize(inferText));
  const candidates = [];
  for (const entity of mergeBioTokens(entities, folded, indexOffsets)) {
    const label = GROUP_TO_LABEL[entity.group.toUpperCase()];
    if (label === undefined)
      continue;
    if (entity.score < EXTEND_SCORE || entity.end <= entity.start)
      continue;
    candidates.push({
      start: entity.start,
      end: entity.end,
      label,
      score: entity.score,
      source: "ner",
      text: raw.slice(entity.start, entity.end)
    });
  }
  return repairSpans(raw, candidates, minScore);
}
function stripBio(label) {
  const m = /^([BI])-(.+)$/.exec(label);
  return m ? { prefix: m[1], base: m[2] } : { prefix: null, base: label };
}
function foldForModel(raw) {
  let text = "";
  const rawStart = [];
  const rawEnd = [];
  let i = 0;
  for (const codePoint of raw) {
    const folded = codePoint.toLowerCase().normalize("NFKD").replace(COMBINING_MARKS_RE, "");
    for (const ch of folded) {
      text += ch;
      rawStart.push(i);
      rawEnd.push(i + codePoint.length);
    }
    i += codePoint.length;
  }
  return { text, rawStart, rawEnd };
}
function buildTokenIndexOffsets(folded, contentTokens) {
  const offsets = [[0, 0]];
  let cursor = 0;
  for (const token of contentTokens) {
    const piece = token.startsWith("##") ? token.slice(2) : token;
    const at = token.startsWith("##") ? cursor : folded.text.indexOf(piece, cursor);
    if (piece.length === 0 || at < 0 || at + piece.length > folded.text.length) {
      offsets.push([0, 0]);
      continue;
    }
    offsets.push([folded.rawStart[at], folded.rawEnd[at + piece.length - 1]]);
    cursor = at + piece.length;
  }
  offsets.push([0, 0]);
  return offsets;
}
function offsetFromTokenIndex(indexOffsets, index) {
  if (indexOffsets === undefined || index === undefined)
    return null;
  const pair = indexOffsets[index];
  if (pair === undefined || pair[0] === pair[1])
    return null;
  return pair;
}
function mergeBioTokens(entities, folded, indexOffsets) {
  const out = [];
  let cursor = 0;
  let current = null;
  const flush = () => {
    if (current !== null) {
      out.push({ group: current.group, score: current.score / current.count, start: current.start, end: current.end });
      current = null;
    }
  };
  for (const entity of entities) {
    if (entity.entity_group !== undefined && typeof entity.start === "number" && typeof entity.end === "number") {
      flush();
      out.push({ group: entity.entity_group, score: entity.score, start: entity.start, end: entity.end });
      continue;
    }
    const rawLabel = entity.entity ?? entity.entity_group;
    if (rawLabel === undefined)
      continue;
    const { prefix, base } = stripBio(rawLabel);
    const indexed = offsetFromTokenIndex(indexOffsets, entity.index);
    let start;
    let end;
    if (indexed !== null) {
      [start, end] = indexed;
    } else {
      const word = (entity.word ?? "").replace(/^##/, "").toLowerCase();
      if (!word)
        continue;
      const at = folded.text.indexOf(word, cursor);
      if (at < 0)
        continue;
      start = folded.rawStart[at];
      end = folded.rawEnd[at + word.length - 1];
      cursor = at + word.length;
    }
    const isSubword = (entity.word ?? "").startsWith("##");
    const continues = current !== null && current.group === base && (prefix !== "B" || isSubword);
    if (continues && current !== null) {
      current.end = end;
      current.score += entity.score;
      current.count += 1;
    } else {
      flush();
      current = { group: base, score: entity.score, start, end, count: 1 };
    }
  }
  flush();
  return out;
}
function repairSpans(raw, spans, anchorScore) {
  let kept = spans.filter((span) => span.score >= anchorScore).map(copySpan);
  const candidates = spans.filter((span) => span.score >= EXTEND_SCORE && span.score < anchorScore).map(copySpan);
  const MAX_ITERS = 32;
  let iters = 0;
  let changed = true;
  while (changed && iters < MAX_ITERS) {
    changed = false;
    iters++;
    for (let i = candidates.length - 1;i >= 0; i--) {
      const candidate = candidates[i];
      if (kept.some((span) => canBridge(raw, candidate, span))) {
        kept.push(candidate);
        candidates.splice(i, 1);
        changed = true;
      }
    }
    const merged = mergeAdjacentConnectors(raw, kept);
    const didMerge = merged.length !== kept.length || merged.some((span, index) => span.start !== kept[index]?.start || span.end !== kept[index]?.end);
    if (didMerge) {
      changed = true;
    }
    kept = merged;
    for (let i = 0;i < kept.length; i++) {
      const repaired = rescueCapitalizedParticles(raw, kept[i], kept, i);
      if (repaired.start !== kept[i].start || repaired.end !== kept[i].end) {
        kept[i] = repaired;
        changed = true;
      }
    }
  }
  return kept.map((span) => ({ ...span, text: raw.slice(span.start, span.end) })).sort((a, b) => a.start - b.start || a.end - b.end);
}
function copySpan(span) {
  return { ...span };
}
function isInitialChar(raw, idx) {
  const c = raw[idx];
  if (c === undefined || !/\p{Lu}/u.test(c))
    return false;
  const prev = raw[idx - 1];
  return prev === undefined || !/\p{L}/u.test(prev);
}
function canBridge(raw, a, b) {
  if (a.label !== b.label)
    return false;
  const [left, right] = a.start <= b.start ? [a, b] : [b, a];
  const gap = raw.slice(left.end, right.start);
  if (!CONNECTOR_RE.test(gap))
    return false;
  if (gap.includes(".") && !isInitialChar(raw, left.end - 1))
    return false;
  return true;
}
function mergeAdjacentConnectors(raw, spans) {
  const merged = [];
  for (const span of [...spans].sort((a, b) => a.start - b.start || a.end - b.end)) {
    const previous = merged[merged.length - 1];
    if (previous !== undefined && canBridge(raw, previous, span)) {
      merged[merged.length - 1] = {
        ...previous,
        end: Math.max(previous.end, span.end),
        score: Math.max(previous.score, span.score),
        text: raw.slice(previous.start, Math.max(previous.end, span.end))
      };
    } else {
      merged.push(copySpan(span));
    }
  }
  return merged;
}
function rescueCapitalizedParticles(raw, span, all = [], selfIndex = -1) {
  if (!PERSON_LABELS.has(span.label))
    return span;
  let leftBound = 0;
  let rightBound = raw.length;
  for (let i = 0;i < all.length; i++) {
    if (i === selfIndex)
      continue;
    const other = all[i];
    if (other.end <= span.start && other.end > leftBound)
      leftBound = other.end;
    if (other.start >= span.end && other.start < rightBound)
      rightBound = other.start;
  }
  let start = span.start;
  let end = span.end;
  const left = LEFT_PARTICLE_RE.exec(raw.slice(0, start));
  if (left !== null && CONNECTOR_RE.test(left[2]) && (!left[2].includes(".") || left[1].length === 1) && start - left[0].length >= leftBound) {
    start -= left[0].length;
  }
  const right = RIGHT_PARTICLE_RE.exec(raw.slice(end));
  if (right !== null && !right[1].includes(".") && end + right[0].length <= rightBound) {
    end += right[0].length;
  }
  return start === span.start && end === span.end ? span : { ...span, start, end, text: raw.slice(start, end) };
}

// src/ner/worker.ts
function registerNerWorker(scope) {
  let classifierPromise = null;
  scope.onmessage = async (event) => {
    const message = event.data;
    if (message.kind === "init") {
      try {
        classifierPromise = loadNerClassifier(message.options);
        await classifierPromise;
        scope.postMessage({ kind: "ready" });
      } catch (error) {
        classifierPromise = null;
        scope.postMessage({ kind: "error", message: String(error) });
      }
      return;
    }
    if (message.kind === "detect") {
      try {
        if (classifierPromise === null)
          throw new Error("[pii-filter] worker not initialized");
        const classifier = await classifierPromise;
        const spans = await detectNer(message.text, classifier, message.minScore);
        scope.postMessage({ kind: "result", id: message.id, spans });
      } catch (error) {
        scope.postMessage({ kind: "error", id: message.id, message: String(error) });
      }
    }
  };
}
function createWorkerClassifier(worker, options) {
  let nextId = 0;
  const pending = new Map;
  let resolveReady = () => {};
  let rejectReady = () => {};
  const ready = new Promise((resolve, reject) => {
    resolveReady = resolve;
    rejectReady = reject;
  });
  worker.onmessage = (event) => {
    const data = event.data;
    if (data.kind === "ready") {
      resolveReady();
      return;
    }
    if (data.kind === "error" && data.id === undefined) {
      rejectReady(new Error(data.message ?? "[pii-filter] worker init failed"));
      return;
    }
    if (data.id === undefined)
      return;
    const entry = pending.get(data.id);
    if (entry === undefined)
      return;
    pending.delete(data.id);
    if (data.kind === "error")
      entry.reject(new Error(data.message));
    else
      entry.resolve(data.spans);
  };
  worker.postMessage({ kind: "init", options });
  function detect(text, minScore) {
    const id = nextId++;
    return new Promise((resolve, reject) => {
      pending.set(id, { resolve, reject });
      worker.postMessage({ kind: "detect", id, text, minScore });
    });
  }
  return { ready, detect };
}

// src/validators.ts
function isLuhnValid(digits) {
  let sum = 0;
  let double = false;
  for (let i = digits.length - 1;i >= 0; i--) {
    let d = digits.charCodeAt(i) - 48;
    if (double) {
      d *= 2;
      if (d > 9)
        d -= 9;
    }
    sum += d;
    double = !double;
  }
  return sum % 10 === 0;
}
function isValidSsn(digits) {
  if (digits.length !== 9)
    return false;
  const area = digits.slice(0, 3);
  const group = digits.slice(3, 5);
  const serial = digits.slice(5);
  if (area === "000" || area === "666")
    return false;
  if (Number(area) >= 900)
    return false;
  if (group === "00")
    return false;
  if (serial === "0000")
    return false;
  return true;
}

// src/heuristics.ts
var DIGIT_RULES = [
  { label: "CREDIT_CARD", lengths: [16, 15, 14], validate: isLuhnValid },
  { label: "SSN", lengths: [9], validate: isValidSsn }
];
var DIGIT_RUN = /\d(?:[ .-]?\d)*/g;
function extractRuns(raw) {
  const runs = [];
  DIGIT_RUN.lastIndex = 0;
  for (let m = DIGIT_RUN.exec(raw);m !== null; m = DIGIT_RUN.exec(raw)) {
    const digits = [];
    const rawIndex = [];
    for (let i = 0;i < m[0].length; i++) {
      const code = m[0].charCodeAt(i);
      if (code >= 48 && code <= 57) {
        digits.push(m[0][i]);
        rawIndex.push(m.index + i);
      }
    }
    runs.push({ digits: digits.join(""), rawIndex });
  }
  return runs;
}
function detectDigitEntities(raw) {
  const spans = [];
  for (const run of extractRuns(raw)) {
    for (const rule of DIGIT_RULES) {
      if (!rule.lengths.includes(run.digits.length))
        continue;
      if (!rule.validate(run.digits))
        continue;
      const start = run.rawIndex[0];
      const end = run.rawIndex[run.rawIndex.length - 1] + 1;
      spans.push({ start, end, label: rule.label, score: 1, source: "heuristic", text: raw.slice(start, end) });
      break;
    }
  }
  return spans;
}
var TEXT_RULES = [
  { label: "EMAIL", pattern: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g },
  { label: "URL", pattern: /\bhttps?:\/\/[^\s<>"'\])}]+/g },
  { label: "URL", pattern: /\bwww\.[A-Za-z0-9.-]+\.[A-Za-z]{2,}(?:\/[^\s<>"'\])}]*)?/g },
  { label: "IP_ADDRESS", pattern: /\b(?:(?:25[0-5]|2[0-4]\d|1?\d?\d)\.){3}(?:25[0-5]|2[0-4]\d|1?\d?\d)\b/g },
  {
    label: "IP_ADDRESS",
    pattern: /(?<![:.\w])(?:(?:[0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,7}:|(?:[0-9A-Fa-f]{1,4}:){1,6}:[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,5}(?::[0-9A-Fa-f]{1,4}){1,2}|(?:[0-9A-Fa-f]{1,4}:){1,4}(?::[0-9A-Fa-f]{1,4}){1,3}|(?:[0-9A-Fa-f]{1,4}:){1,3}(?::[0-9A-Fa-f]{1,4}){1,4}|(?:[0-9A-Fa-f]{1,4}:){1,2}(?::[0-9A-Fa-f]{1,4}){1,5}|[0-9A-Fa-f]{1,4}:(?::[0-9A-Fa-f]{1,4}){1,6}|::(?:[0-9A-Fa-f]{1,4}:){0,6}[0-9A-Fa-f]{1,4})(?![:.\w])/g
  },
  { label: "IP_ADDRESS", pattern: /\b(?:[0-9A-Fa-f]{2}[:-]){5}[0-9A-Fa-f]{2}\b/g }
];
function detectTextEntities(raw) {
  const spans = [];
  for (const { label, pattern, group } of TEXT_RULES) {
    pattern.lastIndex = 0;
    for (let m = pattern.exec(raw);m !== null; m = pattern.exec(raw)) {
      const text = group === undefined ? m[0] : m[group];
      const offset = group === undefined ? 0 : m[0].indexOf(text);
      spans.push({
        start: m.index + offset,
        end: m.index + offset + text.length,
        label,
        score: 1,
        source: "heuristic",
        text
      });
    }
  }
  return spans;
}
function detectHeuristics(raw) {
  return [
    ...detectDigitEntities(raw),
    ...detectTextEntities(raw)
  ];
}

// src/premask.ts
function sentinelFor(label) {
  return `[${label}]`;
}
function premask(raw, spans) {
  const ordered = mergeSpans(spans).slice().sort((a, b) => a.start - b.start);
  let masked = "";
  const rawStart = [];
  const rawEnd = [];
  const copyVerbatim = (from, to) => {
    for (let i = from;i < to; i++) {
      masked += raw[i];
      rawStart.push(i);
      rawEnd.push(i + 1);
    }
  };
  let cursor = 0;
  for (const span of ordered) {
    if (span.start < cursor)
      continue;
    copyVerbatim(cursor, span.start);
    const sentinel = sentinelFor(span.label);
    for (const ch of sentinel) {
      masked += ch;
      rawStart.push(span.start);
      rawEnd.push(span.end);
    }
    cursor = span.end;
  }
  copyVerbatim(cursor, raw.length);
  return { masked, rawStart, rawEnd };
}
function projectMaskedSpan(span, raw, map) {
  if (span.end <= span.start)
    return null;
  const start = map.rawStart[span.start];
  const end = map.rawEnd[span.end - 1];
  if (start === undefined || end === undefined || end <= start)
    return null;
  return { ...span, start, end, text: raw.slice(start, end) };
}

// src/session.ts
var PLACEHOLDER_PATTERN = /\[[A-Z][A-Z_]*_\d+\]/g;

class SessionEntityTable {
  aliases;
  keepLabels;
  forward = new Map;
  reverse = new Map;
  counters = new Map;
  constructor(aliases = {}, keepLabels = KEEP_LABELS) {
    this.aliases = aliases;
    this.keepLabels = keepLabels;
  }
  displayName(label) {
    return this.aliases[label] ?? label;
  }
  placeholderFor(label, value) {
    const key = `${label}:${value.toLowerCase().replace(/\s+/g, " ").trim()}`;
    const existing = this.forward.get(key);
    if (existing !== undefined)
      return existing;
    const name = this.displayName(label);
    const next = (this.counters.get(name) ?? 0) + 1;
    this.counters.set(name, next);
    const token = `[${name}_${next}]`;
    this.forward.set(key, token);
    this.reverse.set(token, value);
    return token;
  }
  scrub(raw, spans) {
    const redactable = applyPolicy(spans, this.keepLabels);
    const placeholders = [];
    let text = raw;
    for (const span of redactable) {
      const token = this.placeholderFor(span.label, span.text);
      placeholders.push(token);
      text = `${text.slice(0, span.start)}${token}${text.slice(span.end)}`;
    }
    return { text, placeholders: placeholders.reverse() };
  }
  rehydrate(text) {
    return text.replace(PLACEHOLDER_PATTERN, (token) => this.reverse.get(token) ?? token);
  }
  knows(token) {
    return this.reverse.has(token);
  }
}

// src/streaming.ts
var PARTIAL_TOKEN = /\[[A-Z_]*(?:_\d*)?$/;

class StreamingReveal {
  resolve;
  buffer = "";
  constructor(resolve) {
    this.resolve = resolve;
  }
  push(chunk) {
    this.buffer += chunk;
    const revealed = this.replaceComplete(this.buffer);
    const partial = revealed.match(PARTIAL_TOKEN);
    if (partial === null) {
      this.buffer = "";
      return revealed;
    }
    const cut = revealed.length - partial[0].length;
    this.buffer = revealed.slice(cut);
    return revealed.slice(0, cut);
  }
  flush() {
    const out = this.replaceComplete(this.buffer);
    this.buffer = "";
    return out;
  }
  replaceComplete(text) {
    return text.replace(PLACEHOLDER_PATTERN, (token) => this.resolve(token) ?? token);
  }
}
function createRevealTransform(resolve) {
  const reveal = new StreamingReveal(resolve);
  return new TransformStream({
    transform(chunk, controller) {
      const out = reveal.push(chunk);
      if (out)
        controller.enqueue(out);
    },
    flush(controller) {
      const out = reveal.flush();
      if (out)
        controller.enqueue(out);
    }
  });
}

// src/guard.ts
var DEFAULT_ALIASES = {};

class ChatGuard {
  table;
  ner;
  noPrefilter;
  constructor(config = {}) {
    this.table = new SessionEntityTable(config.aliases, resolveKeepLabels(config.keepLabels));
    this.ner = config.ner;
    this.noPrefilter = config.noPrefilter ?? false;
  }
  async detect(text) {
    const heuristic = detectHeuristics(text);
    if (this.ner === undefined)
      return heuristic;
    if (this.noPrefilter) {
      const modelSpans = await this.ner(text);
      return [...heuristic, ...modelSpans];
    }
    const map = premask(text, heuristic);
    const maskedSpans = await this.ner(map.masked);
    const contextual = [];
    for (const span of maskedSpans) {
      const projected = projectMaskedSpan(span, text, map);
      if (projected !== null)
        contextual.push(projected);
    }
    return [...heuristic, ...contextual];
  }
  async protect(text) {
    const spans = await this.detect(text);
    return this.table.scrub(text, spans);
  }
  reveal(reply) {
    return this.table.rehydrate(reply);
  }
  revealTransform() {
    return createRevealTransform((token) => {
      const restored = this.table.rehydrate(token);
      return restored === token ? null : restored;
    });
  }
  async protectReply(reply) {
    const spans = await this.detect(reply);
    return this.table.scrub(reply, spans);
  }
}
async function buildNer(options) {
  const { model, worker, minScore, device = "wasm" } = options;
  const modelOptions = { model, device, minScore };
  if (worker !== undefined) {
    const port = new Worker(worker, { type: "module" });
    const classifier2 = createWorkerClassifier(port, modelOptions);
    await classifier2.ready;
    return async (text) => await classifier2.detect(text);
  }
  const classifier = await loadNerClassifier(modelOptions);
  return (text) => detectNer(text, classifier, minScore);
}
async function createGuard(options = {}) {
  const { aliases = DEFAULT_ALIASES, keepLabels, noPrefilter, ner, heuristicsOnly, ...nerLoad } = options;
  let detector = ner;
  if (detector === undefined && heuristicsOnly !== true) {
    detector = await buildNer(nerLoad);
  }
  return new ChatGuard({ ner: detector, aliases, keepLabels, noPrefilter });
}
export {
  shouldRedact,
  sentinelFor,
  resolveKeepLabels,
  registerNerWorker,
  projectMaskedSpan,
  premask,
  mergeSpans,
  loadNerClassifier,
  isValidSsn,
  isLuhnValid,
  detectNer,
  detectHeuristics,
  createWorkerClassifier,
  createRevealTransform,
  createGuard,
  applyPolicy,
  StreamingReveal,
  SessionEntityTable,
  RAMPART_MODEL_ID,
  PLACEHOLDER_PATTERN,
  NER_TOKEN_OVERLAP,
  NER_TOKEN_BUDGET,
  KEEP_LABELS,
  DEFAULT_ALIASES,
  ChatGuard
};

//# debugId=E09D0FA75BFC6A1964756E2164756E21
